src/flask_smorest/exceptions.py

```diff
@@ -7,30 +7,166 @@ class FlaskSmorestError(Exception):
     """Generic flask-smorest exception"""
 
 
+# Non-API exceptions
 class MissingAPIParameterError(FlaskSmorestError):
     """Missing API parameter"""
 
 
-class NotModified(wexc.HTTPException, FlaskSmorestError):
-    """Resource was not modified (Etag is unchanged)
+class CurrentApiNotAvailableError(FlaskSmorestError):
+    """`current_api` not available"""
 
-    Exception created to compensate for a lack in Werkzeug (and Flask)
-    """
 
-    code = 304
-    description = "Resource not modified since last request."
+# API exceptions
+class ApiException(wexc.HTTPException, FlaskSmorestError):
```

Comment:
Wondering it it needs to inherit `HTTPException`. Perhaps. No objection, just wondering.

Reply:
```python
class NotModified(wexc.HTTPException, FlaskSmorestError):
    """Resource was not modified (Etag is unchanged)

    Exception created to compensate for a lack in Werkzeug (and Flask)
    """

    code = 304
    description = "Resource not modified since last request."


class PreconditionRequired(wexc.PreconditionRequired, FlaskSmorestError):
    """Etag required but missing"""

    # Overriding description as we don't provide If-Unmodified-Since
    description = 'This request is required to be conditional; try using "If-Match".'


class PreconditionFailed(wexc.PreconditionFailed, FlaskSmorestError):
    """Etag required and wrong ETag provided"""

```

This is part of original exceptions.py for flask_smorest. As you can see here, flask_smorest inherits and overrides a few HttpExceptions, so I wanted to make sure that I was calling them "http API errors", so I overridden them. (This is good, in my opinion, because it makes it clear what are API errors and what are internal errors that are not).

Reply:
```python
class PreconditionRequired(wexc.PreconditionRequired, ApiException):
    """Etag required but missing"""

    # Overriding description as we don't provide If-Unmodified-Since
    description = 'This request is required to be conditional; try using "If-Match".'


class NotModified(ApiException):
    """Resource was not modified (Etag is unchanged)

    Exception created to compensate for a lack in Werkzeug (and Flask)
    """

    code = 304
    description = "Resource not modified since last request."


for code, exc in wexc.default_exceptions.items():
    api_exc = type(exc.__name__, (exc, ApiException), {})
    setattr(module, exc.__name__, api_exc)

```

As you said, with this approach, we could subclass all Werkzeug errors to our exception class. 

![image](https://github.com/marshmallow-code/flask-smorest/assets/88619089/803e9a15-5cdb-48ca-bb73-90de1a104911)

I'm a little curious here. Should we also allow users to import and use syntax like `from flask_smorest.exceptions import NotFound`? Because it doesn't seem to work when importing in code, not in the REPL.

![image](https://github.com/marshmallow-code/flask-smorest/assets/88619089/504f44cf-50d6-4e93-b1de-e0fceb91fe15)

For example, what code should a user use to get a 404 error?

```python

# 1
from flask_smorest.exceptions import ApiException
raise ApiException.get_exception_by_code(404)(...)

# 2
from flask_smorest imort abort
abort(404)

# 3
from flask_smorest.exceptions import NotFound
raise NotFound(...)

#4 
# maybe.. is this possible?
from flask_smorest.exceptions import ApiException
raise ApiException(404)
```

Would love to hear your thoughts :)